pointcloud/octree/basic.go

```diff
@@ -0,0 +1,171 @@
+package octree
+
+import (
+	"context"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"github.com/pkg/errors"
+
+	pc "go.viam.com/rdk/pointcloud"
+)
+
+// octree data is the structure of an octree.
+type basicOctree struct {
+	ctx    context.Context
+	logger golog.Logger
+	node   basicOctreeNode
+	center r3.Vector
+	side   float64
+	meta   MetaData
+}
+
+type basicOctreeNode struct {
+	nodeType NodeType
+	tree     []*basicOctree
+	point    pc.PointAndData
+}
+
+// New creates a new basic octree with specified center, side and metadata.
+func New(ctx context.Context, center r3.Vector, sideLength float64, logger golog.Logger) (Octree, error) {
+	if sideLength <= 0 {
+		return nil, errors.Errorf("invalid side length (%.2f) for octree", sideLength)
+	}
+
+	octree := &basicOctree{
+		ctx:    ctx,
+		logger: logger,
+		node:   newLeafNodeEmpty(),
+		center: center,
+		side:   sideLength,
+	}
+	octree.meta = octree.NewMetaData()
+	return octree, nil
+}
+
+// Size returns the number of points stored in the octree by traversing it and incrementing based on the number
+// of nodes containing a point.
+func (octree *basicOctree) Size() int {
```

Comment:
Could this number be cached? It feels weird needing to do an O(N) calculation every time you need to know how big the octree is (not to mention you are putting a TON of Size() calls on the stack since its recursive). This function will take way longer to execute than the At() function, and maybe just as long as the Iterate function.

You should save size as a struct variable in the octree, and store the tree's size in it

Reply:
This num is cached, and is inside of metadata. I most access of the size variable will be done through that and will change Size to accessing that quantity instead.

Reply:
Fixed by including in node struct